### PR TITLE
replace the old building and packaging tools with poetry

### PR DESCRIPTION
**Pull request by HosamAlmoghraby**
- With the changes here I am adding poetry as dependencies and packaging management tool to cookiecutter. as setuptools and wheels are not needed anymore, they were removed from the ci workflows.

- The publishing workflow is tested by building and pushing the package to the test repository of PyPI. here is the test release https://test.pypi.org/project/cookiecutter/2.0.2/

- The **CI Tests** workflows are successful on all python versions.

### CODE REVIEW

File: pyproject.toml
Comment (browniebroke) [inline pyproject.toml:84]: This is breaking the `tool` table into what is called an "out-of-order" tables. The [TOML spec discourages it](https://toml.io/en/v1.0.0#table):

> Defining tables out-of-order is discouraged.
> 
> ```toml
> # VALID BUT DISCOURAGED
> [fruit.apple]
> [animal]
> [fruit.orange]
> ```
>
> ```toml
> # RECOMMENDED
> [fruit.apple]
> [fruit.orange]
> [animal]
> ```

Personally, I've hit some nastly bugs with toiml parsers in the past with out-of-order tables. I would strongly recommend moving this table outside of the `tool` table.

**Review by browniebroke — COMMENTED** [review]
Not a cookicutter maintainer, but I have some experience with Poetry, I've noted a couple of things you might find useful.

File: pyproject.toml
Comment (HosamAlmoghraby) [reply pyproject.toml]: thank you @browniebroke

File: pyproject.toml
Comment (HosamAlmoghraby) [reply pyproject.toml]: thank you

File: pyproject.toml
Comment (HosamAlmoghraby) [inline pyproject.toml:14]: add maintainers meta

File: pyproject.toml
Comment (pydanny) [inline pyproject.toml:12]:
```suggestion
    "Audrey Roy Greenfeld <audrey@feldroy.com>"
```

**Review by pydanny — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (pydanny) [inline pyproject.toml:45]:
```suggestion
```

**Review by pydanny — CHANGES_REQUESTED** [review]
Add `release-process.md` that indexes and describes each file in the release process.

File: tox.ini
Comment (HosamAlmoghraby) [inline tox.ini:36]: install dev requirements from pyproject.toml file

File: .github/workflows/docs.yml
Comment (browniebroke) [inline .github/workflows/docs.yml:30]: This can be simplified once `actions/setup-python` supports it natively: https://github.com/actions/setup-python/pull/281

File: .github/workflows/docs.yml
Comment (browniebroke) [inline .github/workflows/docs.yml:21]: Might want to consider using a [dedicated GH action](https://github.com/snok/install-poetry) for that:

```yaml
- name: Install and configure Poetry
  uses: snok/install-poetry@v1
  with:
    virtualenvs-in-project: false
```

Unless there is a good reason not to?

File: tox.ini
Comment (browniebroke) [inline tox.ini:34]: Just noting that Poetry will add the concept of [Dependency groups in 1.2](https://python-poetry.org/blog/announcing-poetry-1.2.0a2/#dependency-groups) which might be a good place to track documentation dependencies.

I've been watching this for a while but sadly it's still in alpha...

**Review by browniebroke — COMMENTED** [review]
Disclaimer (again): not a maintainer, just a fellow user interested in helping, so take these with a grain of salt 😄

File: tox.ini
Comment (HosamAlmoghraby) [reply tox.ini:32]: this is nice to have..  thank you for sharing...

File: .github/workflows/docs.yml
Comment (HosamAlmoghraby) [reply .github/workflows/docs.yml:30]: do you think this will be included in the setup-python action v3 or this is a minor change?

File: .github/workflows/docs.yml
Comment (HosamAlmoghraby) [reply .github/workflows/docs.yml:21]: I have already discussed that in my last meeting with Audrey and Daniel, we might use this packed action in a separate pull request, but for now, we will go with these multi-steps workflow.
btw, I already tried this action from SNOK and it seems it worked fine.

File: .github/workflows/docs.yml
Comment (browniebroke) [reply .github/workflows/docs.yml:30]: They've added caching as non-breaking change, and this feels smaller and non-breaking as well, so to could be in v2, unless there is an unrelated breaking change before it's merged, of course.

File: release-process.md
Comment (insspb) [inline release-process.md:2]: Blank line missed after header. Will fail on generating in rd

File: tox.ini
Comment (insspb) [inline tox.ini:39]: Blank line at the end of file missed

**Review by insspb — CHANGES_REQUESTED** [review]
I am not sure about poetry benefits, but at least errors should be fixed.

**Comment by insspb** [comment]
Squash on merge.

File: release-process.md
Comment (HosamAlmoghraby) [reply release-process.md:3]: the missing blank lines has been added

File: tox.ini
Comment (HosamAlmoghraby) [reply tox.ini]: blank line added to the end of the file...

**Comment by jensens** [comment]
With 2.1.0 we follow now the official PyPA process for Github actions  as defined here: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/

As Cookiecutter defines itself as a conservative project (https://cookiecutter.readthedocs.io/en/latest/CONTRIBUTING.html#implement-features) I tend to propose not merge this PR, since following the official way is IMO the conservative way.

I personally doubt we have any advantages with Poetry here. But decision is up to the @cookiecutter/cookiecutter-sentinel core contributor team and the leads as a whole.

What do we gain using Poetry? 
What problems does it solve and what are the advantages over the conservative way?

**Comment by ssbarnea** [comment]
I also vote against poetry. Setuptools and PEP-517 are mature and a safe bet as of today.

**Comment by HosamAlmoghraby** [comment]
The work here started 7 months and was waiting for a merge since January. If I start now arguing how Poetry facilitates the packaging and whether it brings advantages to the project or not and how it solves problems, then I feel something is going wrong here with the timing.

So as @jensens and @ericof are the new maintainers of Cookiecutter, here is a brief history of this PR:
Replacing the setuptools with Poetry was already discussed during the live mob programming sessions last December with @audreyfeldroy when everybody was struggling to figure out how to come up with a new release because of the lack of documentation. Back then in the Christmas holidays, I came to an agreement with @audreyfeldroy that I go deeper into how the workflow in Cookiecutter is built and I suggested leveraging the packaging to Poetry.
We agreed on a meeting on 14th January where I showed the changes I made to @pydanny and @audreyfeldroy and I also explained how Poetry brings advantages to Cookiecutter.
After that meeting, I had a few previews and I made some changes as agreed during the meeting.
Also on the same day of that meeting, I was able to publish version 2.0.3 to the test PyPI registry using the Poetry and many volunteers participated in a community testing.
Since then, we agreed several times on a meeting where we publish the new release with the new workflow but @pydanny was not able to show up because of private reasons.

**Comment by jensens** [comment]
@HosamAlmoghraby let me try to summarize: 
- the work started because of lack of knowledge how to implement a proper release process, 
- Poetry seemed to be a possible solution to overcome this
- there were some discussion, but its not documented nor are there any facts left to built upon

Meanwhile @ericof and I implemented the official PyPA release workflow (it was like 2h work). So there is no reason anymore left from what I get to merge this. It hurts, because work and time flew into this PR, but OTHO this how evolution of software works. I feel the pain with you.

Question is - from the current state of 2.1.0: What are the benefits? What do we gain? I do not see much. One thing I like is there is no `setup.py` anymore. But this is possible with the official tools as well.

As said, I am conservative here. Unless I do not find a real problem solved, it is not worth to merge. Just for the sake of Poetry is not a reason.

**Comment by HosamAlmoghraby** [comment]
- What are the benefits? What do we gain?
From the current state, I think, not so much (talking from a conservative point of view). The current implementation of the workflow and the other workflow proposed in this PR, both lead to package and publish Cookiecutter to PyPI.
Poetry is an extra tool for dependency management and packaging. It manages this process of the box, It does that like any other library used in the project to do a specific job. (click for example is used as a command-line interface in Cookiecutter!).

As long as the core team voted against the changes proposed here, I am fine with closing this PR.

**Pull request closed without merging**
